MAINT: Move np.char methods to np.strings

lysnikolaou opened the pull request:
- Remove all implementations from `np.char` and adds them in `np.strings`
- Reexports them from `np.char` for backwards-compatibility.
- Does not touch tests, but since `test_defchararray.py` passes, it should all be okay.
- Will improve docstrings as I open the ufunc PRs.

Closes #25567.
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

lysnikolaou committed:
MAINT: Move np.char methods to np.strings

lysnikolaou committed:
Fix tests that verify that methods are exposed from one place

ngoldbaum commented:
It looks like the docs for `np.char` and `np.strings` are all looking good in the docs build: 

https://output.circle-artifacts.com/output/job/4e8fc962-9a4a-4db2-a4da-c09446de313e/artifacts/0/doc/build/html/reference/routines.char.html

https://output.circle-artifacts.com/output/job/4e8fc962-9a4a-4db2-a4da-c09446de313e/artifacts/0/doc/build/html/reference/routines.strings.html

Nice!

The duplication isn't great though, and we should do a little more to reduce user confusion arising from that in the docs. How about we split off the docs for `np.char` in the table of contents for the API reference into a new "Legacy Routines" section below "Routines"? 

Can you please also add the `.. legacy::` directive to the header for the `np.char` docs, along with a note urging to use string arrays and the functions in `np.strings`? See the `howto-docs` page in the dev docs for more detail about that, it looks like we'll be the first user outside of that example of how to use the directive, but I bet it will get a lot more use as we do a once-over on the docs before the final 2.0 release.

ngoldbaum commented on numpy/_core/defchararray.py:
```diff
@@ -18,13 +18,11 @@
 from numpy._core import overrides
-from numpy._utils import asbytes
-import numpy
+from numpy.strings import *
```
Note to other reviewers, we do use `import *` heavily in the private API and using it here is fine.

ngoldbaum commented on numpy/_core/defchararray.py:
```diff
@@ -2690,7 +840,7 @@ def partition(self, sep):
         return asarray(partition(self, sep))
 
-    def replace(self, old, new, count=None):
+    def replace(self, old, new, count=-1):
```
I think probably best to leave the default alone in a member function of `chararray` because it's effectively deprecated and is poorly tested. I think the `_check_args` function in `np._core.strings` handle dropping the `None` argument.

mhvk replied:
Agreed but for a different reason: this PR should really only move code, not make any user-visible API changes. I think it *is* OK to do `replace(self, old, new, count if count is not None else -1)` below, though! 

p.s. One could also adjust `strings.replace` to allow `count=None` again (like it does in the function removed here), but I think that is less good, since the new definition, with its default of `-1`, matches the python `str.replace` default.

mhvk reviewed (commented):
Looks good! Only one comment...

lysnikolaou committed:
Move np.char docs to legacy section & add np.strings to refguide check

lysnikolaou committed:
Merge branch 'main' into move-np-char

lysnikolaou committed:
Address feedback; revert change to replace default value for count

lysnikolaou commented:
@ngoldbaum @mhvk Sorry for the delay on this. I was out sick for 1,5 weeks.

I've addressed the feedback. Docs look good and the one CI failure is unrelated.

ngoldbaum commented on doc/source/reference/legacy_routines.rst:
```diff
@@ -0,0 +1,24 @@
+.. _legacy_routines:
```
@rossbar, @Mukulikaa, @rgommers - are you ok with this new top-level reference docs page for legacy API? Organizing it this way makes sense to me but want to make sure other people working on similar efforts are in sync on this.

ngoldbaum commented on numpy/_core/strings.py:
```diff
@@ -31,16 +32,272 @@
-    "lstrip", "rstrip", "strip", "replace"
+    "lstrip", "rstrip", "strip", "replace",
+
+    # _vec_string
```
Maybe add a line to this comment saying the plan is to move all these to ufuncs.

ngoldbaum reviewed (approved):
LGTM but I'd like another reviewer who is working on docs for numpy 2.0 to look at this to make sure the new legacy docs page is OK.

lysnikolaou committed:
Merge branch 'main' into move-np-char

lysnikolaou committed:
Improve comment in string.py __all__

rgommers commented:
> LGTM but I'd like another reviewer who is working on docs for numpy 2.0 to look at this to make sure the new legacy docs page is OK.

I'm not a fan of a separate "legacy docs" page, there's no real reason for introducing that, and it's impossible to cover all legacy APIs that way even if we wanted to, since they're not all separate submodules. And the ones that are submodules are already covered by https://numpy.org/devdocs/reference/module_structure.html#legacy-namespaces.

This is not a blocking comment, it's okay with me to merge this and deal with this in a follow-up. But here is what I would do:
- Re-title the `numpy.strings` page from "String universal functions" to "String functionality"
- Re-title the `numpy.char` page from "String operations" to something more specific (e.g., "Legacy fixed-width string functionality")
- On the `numpy.strings` page, add a `.. note::` box mentioning that the legacy API is `np.char` and cross-linking it
- Remove the `routines.char.rst` entry from "Routines and objects by topic" completely

mhvk commented:
For legacy documentation, one could also think of having a super simple page that states that `np.char` is deprecated and providing a link to the 1.26 version of the documentation.

mhvk reviewed (commented):
I've only scanned through the copied part, but looks all fine. The main question indeed seems to be what to do with the deprecation; see my earlier suggestion.

lysnikolaou committed:
Remove legacy pages; add link to np.char from np.strings docs

ngoldbaum commented:
Excellent, `np.char` is still showing up in the API docs but there's no longer a top-level reference doc page for it. Let's bring this in, thanks @lysnikolaou!
